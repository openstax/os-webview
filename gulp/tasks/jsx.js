const path = require('path');
const gulp = require('gulp');
const config = require('../config');
const pi = require('gulp-load-plugins')({
    pattern: ['gulp-*', 'gulp.*', 'del']
});

function jsx() {
    const configSrcApp = `${config.src}/app`;
    const configDestApp = `${config.dest}/app`;

    return gulp.src(`${configSrcApp}/**/*.jsx`, {
        since: gulp.lastRun(jsx)
    })
    .pipe(pi.sourcemaps.init({loadMaps: true}))
    .pipe(pi.rename((uri) => {
        uri.extname = '.jsx.js';
    }))
    .pipe(pi.babel({
        compact: false,
        presets: ['@babel/preset-react']
    }))
    // prefix the sourcemaps with '../src/' so webpack can find them
    .pipe(pi.sourcemaps.mapSources(function(sourcePath, file) {
      const sourcePathHtml = sourcePath.replace('.jsx.js', '.jsx');
      const rel = path.relative(path.dirname(`${configDestApp}/${sourcePath}`), `${configSrcApp}/${sourcePathHtml}`)
      return rel;
    }))
    .pipe(pi.sourcemaps.write('.'))
    .pipe(gulp.dest(`${configDestApp}`));
}


function watchJsx() {
    gulp.watch([`${config.src}/**/*.jsx`], config.watchOpts, gulp.series(
        jsx
    ));
}

exports.jsx = jsx;
exports.jsx.watch = watchJsx;
