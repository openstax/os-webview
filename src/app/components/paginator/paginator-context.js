import {useState, useCallback} from 'react';
import buildContext from '~/components/jsx-helpers/build-context';

function useContextValue(params={}) {
    const {resultsPerPage=1, initialPage=1} = params;
    const [currentPage, setCurrentPage] = useState(initialPage);
    const firstOnPage = (currentPage - 1) * resultsPerPage;
    const lastOnPage = firstOnPage + resultsPerPage - 1;
    const isVisible = useCallback(
        (childIndex) => childIndex >= firstOnPage && childIndex <= lastOnPage,
        [firstOnPage, lastOnPage]
    );
    const visibleChildren = useCallback(
        (children) => children.slice(firstOnPage, lastOnPage + 1),
        [firstOnPage, lastOnPage]
    );

    return {currentPage, setCurrentPage, resultsPerPage, firstOnPage, lastOnPage, isVisible, visibleChildren};
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {
    useContext as default,
    ContextProvider as PaginatorContextProvider
};
