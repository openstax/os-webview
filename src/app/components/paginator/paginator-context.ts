import {useState, useCallback} from 'react';
import buildContext from '~/components/jsx-helpers/build-context';

type ContextParams = {
    resultsPerPage?: number;
    initialPage?: number;
};

function useContextValue(params: ContextParams = {}) {
    const {resultsPerPage = 1, initialPage = 1} = params;
    const [currentPage, setCurrentPage] = useState(initialPage);
    const firstOnPage = (currentPage - 1) * resultsPerPage;
    const lastOnPage = firstOnPage + resultsPerPage - 1;
    const isVisible = useCallback(
        (childIndex: number) =>
            childIndex >= firstOnPage && childIndex <= lastOnPage,
        [firstOnPage, lastOnPage]
    );
    const visibleChildren = useCallback(
        (children: React.ReactNode[]) =>
            children.slice(firstOnPage, lastOnPage + 1),
        [firstOnPage, lastOnPage]
    );

    return {
        currentPage,
        setCurrentPage,
        resultsPerPage,
        firstOnPage,
        lastOnPage,
        isVisible,
        visibleChildren
    };
}

const {useContext, ContextProvider} = buildContext({useContextValue});

export {useContext as default, ContextProvider as PaginatorContextProvider};
