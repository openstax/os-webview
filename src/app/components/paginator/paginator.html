<template args="model">
    <div class="button-bar" if="model.pages > 1">
        <button disabled="{model.disablePrevious}">Previous</button>
        <button each="indicator in model.pageIndicators"
            disabled="{indicator.disabled}"
            aria-selected="{indicator.selected}"
            aria-label="{indicator.page}">{indicator.label}</button>
        <button disabled="{model.disableNext}">Next</button>
    </div>
    <div class="summary">{model.resultRange} of {model.totalResults} for <b>'{model.searchTerm}'</b></div>
</template>
