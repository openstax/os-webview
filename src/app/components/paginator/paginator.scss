@import 'pattern-library/core/pattern-library/headers';

.paginator {
    display: grid;
    justify-content: center;

    @include width-up-to($phone-max) {
        grid-row-gap: 3.7rem;
    }

    @include wider-than($phone-max) {
        grid-row-gap: 4.7rem;
    }

    .button-bar {
        @extend %button-bar;

        > button {
            padding: 0;

            &:not(:first-child):not(:last-child) {
                width: 4.9rem;

                @include width-up-to($phone-max) {
                    width: 3.1rem;
                }
            }

            &:first-child,
            &:last-child {
                width: 11.1rem;

                @include width-up-to($phone-max) {
                    width: 6.7rem;

                    &:last-child {
                        width: 6.5rem;
                    }
                }
            }

            &[aria-selected="true"] {
                background-color: text-color(link);
                color: text-color(white);
            }
        }

    }

    .summary {
        text-align: center;
    }
}
