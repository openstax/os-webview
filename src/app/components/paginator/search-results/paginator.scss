@import 'pattern-library/core/pattern-library/headers';

.paginator {
    display: grid;
    justify-content: center;

    @include width-up-to($phone-max) {
        grid-row-gap: 3.7rem;
    }

    @include wider-than($phone-max) {
        grid-row-gap: 4.7rem;
    }

    .button-bar {
        @extend %button-bar;

        li {
            padding: 0;

            > button {
                background-color: inherit;
                color: inherit;
                height: 100%;
                padding: 0;
                width: 100%;

                &[aria-current="page"] {
                    background-color: text-color(link);
                    color: text-color(white);
                }
            }

            &:not(:first-child):not(:last-child) {
                width: 4.9rem;

                @include width-up-to($phone-max) {
                    width: 3.1rem;
                }
            }

            &:first-child,
            &:last-child {
                width: 11.1rem;

                @include width-up-to($phone-max) {
                    width: 6.7rem;

                    &:last-child {
                        width: 6.5rem;
                    }
                }
            }
        }

    }

    .summary {
        text-align: center;
    }
}
