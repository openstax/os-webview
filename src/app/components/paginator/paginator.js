import React from 'react';
import './paginator.css';

const RESULTS_PER_PAGE = 10;

export function PaginatedResults({currentPage, children}) {
    const firstOnPage = (currentPage - 1) * RESULTS_PER_PAGE;

    return (
        <div className="cards boxed">
            {
                children.slice(firstOnPage, firstOnPage + RESULTS_PER_PAGE)
            }
        </div>
    );
}

function getPageIndicators(pages, currentPage) {
    const indicatorCount = Math.min(pages, 5);
    const propsFor = (label) => ({
        label,
        page: `page ${label}`,
        disabled: label === currentPage,
        selected: label === currentPage
    });
    const result = Array(indicatorCount).fill();

    if (pages - currentPage < 3) {
        result[0] = pages - indicatorCount + 1;
    } else {
        result[0] = (currentPage > 3) ? currentPage - 2 : 1;
    }
    for (let i = 1; i < indicatorCount; ++i) {
        result[i] = result[i-1] + 1;
    }
    return result.map(propsFor);
    return result;
}

function PageButtonBar({currentPage, pages, setCurrentPage}) {
    const disablePrevious = currentPage === 1;
    const disableNext = currentPage === pages;
    const pageIndicators = getPageIndicators(pages, currentPage);

    function prevPage() {
        setCurrentPage(currentPage - 1);
    }
    function nextPage() {
        setCurrentPage(currentPage + 1);
    }

    return (
        <div className="button-bar">
            <button disabled={disablePrevious} onClick={prevPage}>Previous</button>
            {
                pageIndicators.map((indicator) =>
                    <button
                        key={indicator}
                        disabled={indicator.disabled}
                        aria-selected={indicator.selected}
                        aria-label={indicator.page}
                        onClick={() => setCurrentPage(indicator.label)}
                    >{indicator.label}</button>
                )
            }
            <button disabled={disableNext} onClick={nextPage}>Next</button>
        </div>
    );
}

export function PaginatorControls({items, currentPage, setCurrentPage}) {
    const pages = Math.ceil(items / RESULTS_PER_PAGE);
    const firstIndex = (currentPage - 1) * RESULTS_PER_PAGE;
    const endBefore = Math.min(firstIndex + RESULTS_PER_PAGE, items);
    const resultRange = `${firstIndex + 1}-${endBefore}`;
    const searchTerm = decodeURIComponent(window.location.search.substr(1));

    return (
        <div className="paginator">
            {
                pages > 1 &&
                    <PageButtonBar
                        pages={pages}
                        currentPage={currentPage} setCurrentPage={setCurrentPage}
                    />
            }
            <div className="summary">{resultRange} of {items} for <b>'{searchTerm}'</b></div>
        </div>
    );
}
