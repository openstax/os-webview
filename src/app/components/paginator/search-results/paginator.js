import React from 'react';
import usePaginatorContext from '../paginator-context';
import './paginator.scss';

function getPageIndicators(pages, currentPage) {
    const indicatorCount = Math.min(pages, 5);
    const propsFor = (label) => ({
        label,
        page: `page ${label}`,
        disabled: label === currentPage,
        selected: label === currentPage
    });
    const result = Array(indicatorCount).fill();

    if (pages - currentPage < 3) {
        result[0] = pages - indicatorCount + 1;
    } else {
        result[0] = (currentPage > 3) ? currentPage - 2 : 1;
    }
    for (let i = 1; i < indicatorCount; ++i) {
        result[i] = result[i-1] + 1;
    }
    return result.map(propsFor);
}

function PageButtonBar({pages}) {
    const {currentPage, setCurrentPage} = usePaginatorContext();
    const disablePrevious = currentPage === 1;
    const disableNext = currentPage === pages;
    const pageIndicators = getPageIndicators(pages, currentPage);

    function prevPage() {
        setCurrentPage(currentPage - 1);
    }
    function nextPage() {
        setCurrentPage(currentPage + 1);
    }

    return (
        <div className="button-bar" role="listbox">
            <button disabled={disablePrevious} onClick={prevPage}>Previous</button>
            {
                pageIndicators.map((indicator) =>
                    <button
                        role="option"
                        key={indicator}
                        disabled={indicator.disabled}
                        aria-selected={indicator.selected}
                        aria-label={indicator.page}
                        onClick={() => setCurrentPage(indicator.label)}
                    >{indicator.label}</button>
                )
            }
            <button disabled={disableNext} onClick={nextPage}>Next</button>
        </div>
    );
}

export function PaginatorControls({items}) {
    const {currentPage, resultsPerPage} = usePaginatorContext();
    const pages = Math.ceil(items / resultsPerPage);
    const firstIndex = (currentPage - 1) * resultsPerPage;
    const endBefore = Math.min(firstIndex + resultsPerPage, items);
    const resultRange = `${firstIndex + 1}-${endBefore}`;
    const searchTerm = new window.URLSearchParams(window.location.search).get('q');

    return (
        <div className="paginator">
            {pages > 1 && <PageButtonBar pages={pages} />}
            <div className="summary">
                {items > 0 && `${resultRange} of `}
                {items} for <b>&apos;{searchTerm}&apos;</b>
            </div>
        </div>
    );
}
