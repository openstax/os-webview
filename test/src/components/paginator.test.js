import {makeMountRender, snapshotify} from '../../helpers/jsx-test-utils.jsx';
import {PaginatedResults, PaginatorControls} from '~/components/paginator/paginator.jsx';

describe('PaginatorControls', () => {
    let page = 2;
    const wrapper = makeMountRender(PaginatorControls, {
        items: 33,
        currentPage: page,
        setCurrentPage(newPage) {
            page = newPage;
        }
    })();

    it('operates by button clicks', () => {
        expect(snapshotify(wrapper)).toMatchSnapshot();
        const getButtons = () => wrapper.find('button');
        const nextPageButton = getButtons().last();

        expect(nextPageButton.text()).toBe('Next');
        nextPageButton.simulate('click');
        expect(page).toBe(3);
        getButtons().first().simulate('click');
        expect(page).toBe(1);
        getButtons().at(2).simulate('click');
        expect(page).toBe(2);
    });
});
