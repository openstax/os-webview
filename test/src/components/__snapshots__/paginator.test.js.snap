// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatorControls operates by button clicks 1`] = `"<div class=\\"paginator\\"><div class=\\"button-bar\\"><button>Previous</button><button aria-selected=\\"false\\" aria-label=\\"page 1\\">1</button><button disabled=\\"\\" aria-selected=\\"true\\" aria-label=\\"page 2\\">2</button><button aria-selected=\\"false\\" aria-label=\\"page 3\\">3</button><button aria-selected=\\"false\\" aria-label=\\"page 4\\">4</button><button>Next</button></div><div class=\\"summary\\">11-20 of 33 for <b>''</b></div></div>"`;
